XenAdmin/Alerts/Types/LeafCoalesceAlert.cs

```diff
@@ -0,0 +1,114 @@
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE 
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF 
+ * SUCH DAMAGE.
+ */
+
+using XenAPI;
+using static XenAPI.Message;
+
+namespace XenAdmin.Alerts
+{
+    public class LeafCoalesceAlert : MessageAlert
+    {
+        private static readonly log4net.ILog log = log4net.LogManager.GetLogger(System.Reflection.MethodBase.GetCurrentMethod().DeclaringType);
+
+        private readonly VM _vm;
+        private readonly VDI _vDI;
+
+        public LeafCoalesceAlert(Message msg)
+            : base(msg)
+        {
+            var obj = msg.GetXenObject();
+
+            if (obj is VDI vdi && vdi.Connection != null)
+            {
+                foreach (var vbdRef in vdi.VBDs)
+                {
+                    var vbd = vdi.Connection.Resolve(vbdRef);
+
+                    if (vbd != null && vbd.currently_attached)
+                    {
+                        VM vm = vbd.Connection.Resolve(vbd.VM);
+
+                        if (vm != null)
+                        {
+                            _vm = vm;
+                            _vDI = vdi;
+                            break;
+                        }
+                    }
+                }
+            }
+        }
+
+        public override string Description
+        {
+            get
+            {
+                switch (Message.Type)
+                {
+                    case MessageType.LEAF_COALESCE_START_MESSAGE:
+                        return string.Format(Messages.LEAF_COALESCE_START_DESCRIPTION, _vDI.Name(), _vm.Name());
+                    case MessageType.LEAF_COALESCE_COMPLETED:
+                        return string.Format(Messages.LEAF_COALESCE_COMPLETED_DESCRIPTION, _vDI.Name(), _vm.Name());
+                    case MessageType.LEAF_COALESCE_FAILED:
+                        return string.Format(Messages.LEAF_COALESCE_FAILED_DESCRIPTION, _vDI.Name(), _vm.Name());
+                    default:
+                        return base.Description;
+                }
+            }
+        }
+
+        public override string Title
+        {
+            get
+            {
+
+                switch (Message.Type)
+                {
+                    case MessageType.LEAF_COALESCE_START_MESSAGE:
+                        return string.Format(Messages.LEAF_COALESCE_START_TITLE, _vm.Name());
+                    case MessageType.LEAF_COALESCE_COMPLETED:
+                        return string.Format(Messages.LEAF_COALESCE_COMPLETED_TITLE, _vm.Name());
+                    case MessageType.LEAF_COALESCE_FAILED:
+                        return string.Format(Messages.LEAF_COALESCE_FAILED_TITLE, _vm.Name());
+                    default:
+                        return base.Title; ;
+                }
+            }
+        }
+
+        public override string HelpID
+        {
+            get
+            {
+                return "LeafCoalesceAlert";
+            }
+        }

```

Comment:
again very nit picky so only if you're making other changes:


```suggestion
      public override string HelpID => "LeafCoalesceAlert";
```